### PR TITLE
feat: add BatchWorkpool for high-throughput task processing

### PR DESCRIPTION
**Pull request by sethconvex**
## Summary

Adds `BatchWorkpool`, a new execution mode where a small number of long-lived executor actions each run many task handlers concurrently. This eliminates the 1-action-per-task overhead of standard mode and enables much higher effective concurrency.

### Key design decisions

- **Slot-based executor partitioning**: Each task is assigned a random slot (0..maxWorkers-1) on enqueue. Each executor only claims tasks from its slot, eliminating OCC conflicts on the shared pending index — validated with zero OCC errors at 90K tasks.
- **Two-step claiming**: Executors read pending task IDs via a query, then atomically claim them via a mutation with ownership verification, preventing stale executors from interfering.
- **Memory back-pressure**: Executors pause claiming when V8 heap usage exceeds 70%, preventing OOM crashes on large batches.
- **Self-healing watchdog**: A periodic mutation detects missing executors and restarts them, recovering from crashes and redeploys without manual intervention.
- **Claim ownership verification**: `complete`, `fail`, and `release` mutations verify the calling executor still owns the task, preventing duplicate completions from stale executors.

### What's included

- `src/client/batch.ts` — `BatchWorkpool` class with `action()`, `executor()`, `enqueue()`, `enqueueBatch()`, `enqueueByHandle()`
- `src/component/batch.ts` — Component-side mutations: enqueue, claim, complete, fail, release, executor lifecycle
- `src/component/schema.ts` — `batchTasks` table with slot-partitioned indexes, `batchConfig` singleton
- `src/client/batch.test.ts` — 100+ client-side unit tests (executor loop, retry, backoff, memory pressure)
- `src/component/batch.test.ts` — 100+ component-side unit tests (claim ownership, slot isolation, edge cases)
- `example/` — Full comparison app with standard vs batch modes, stress test handlers
- `README.md` — BatchWorkpool documentation and usage guide
- `BATCH_DESIGN_ANALYSIS.md` — Detailed architecture rationale

### Benchmarks: Standard vs Batch mode

Head-to-head comparison on the same Convex dev deployment. Each task runs a 2-second simulated handler (`slowWork`).

| Mode | Tasks | Config | Time | Speedup |
|------|-------|--------|------|---------|
| **Standard** | 1K | maxParallelism=100 | **42s** | 1x |
| **Batch** | 1K | 5 workers × 1000 concurrency | **7s** | **6x** |
| **Standard** | 10K | maxParallelism=100 | **556s** (~9.3 min) | 1x |
| **Batch** | 10K | 5 workers × 1000 concurrency | **20s** | **28x** |

Standard mode processes ~100 tasks every ~2s (limited by maxParallelism and per-action scheduling overhead). Batch mode runs all tasks concurrently across 5 long-lived executors, completing 1K tasks in a single 2s cycle and 10K tasks in ~4 cycles.

### Additional stress tests (batch mode only)

| Test | Tasks | Handler delay | Workers | Time | OCC errors |
|------|-------|--------------|---------|------|------------|
| echo | 10K | 0ms | 5 | 19s | 0 |
| slowWork | 20K | 2s | 5 | 70s | 0 |
| slowWork | 30K | 2s | 5 | 494s | 0 |
| simulatedWork | 90K | 200ms | 5 | 527s | 0 |

Zero OCC errors across all test configurations, validating the slot-based executor partitioning design.

## Test plan

- [x] 217 unit tests pass (`npm test`)
- [x] TypeScript builds clean (`npm run build`)
- [x] Head-to-head benchmark: 6-28x faster than standard mode
- [x] Stress tested on live Convex deployment up to 90K tasks
- [x] Zero OCC errors across all test configurations
- [x] Verified self-healing after executor restarts

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

# Release Notes

* **New Features**
  * Added Batch Execution Mode with BatchWorkpool for high-throughput task processing with memory-aware back-pressure and concurrent batch operations.
  * New batch configuration options for tuning worker concurrency and task retry behavior.
  * Extended API exports to support batch workpool types and operations.

* **Documentation**
  * Added comprehensive Batch Execution Mode guide with design analysis covering architecture, tradeoffs, and performance considerations.

### Latest branch benchmark (2026-02-14)

Additional apples-to-apples operational comparison on this branch:

| Path | Workload | Time |
|------|----------|------|
| Legacy standard path | `example:enqueueABunchOfActions` × 67 (~2010 actions) | **209s** |
| New batch path | `test:runBatch` (`count=2000`, mock translation enabled) | **44s** |

Derived from this run:
- Absolute delta: **165s faster**
- Speedup: **4.75x**
- Percent faster: **78.95%**

Artifact: `benchmark_results/standard_vs_batch_comparison_2026-02-14.txt`

### CODE REVIEW

File: src/client/batch.ts [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+        args: Args,
+      ) => Promise<Returns>;
+    },
+  ): RegisteredAction<"internal", Args, Returns> {
+    this.registry.set(name, opts.handler as HandlerFn);
+    return internalActionGeneric({
+      args: opts.args as any,
+      handler: opts.handler as any,
+    });
+  }
```
[the remark this answers is not in the document]
Comment (sethconvex) [reply src/client/batch.ts]: By design. The `action()` method registers the handler for use by the executor loop while also exposing it as a standalone `internalAction` for direct invocation. The args schema on the standalone action matches what the user passes to `enqueue`, so it does validate.

File: src/component/pool.ts [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+import type { FunctionHandle } from "convex/server";
+import { v } from "convex/values";
+import type { Id } from "./_generated/dataModel.js";
+import { mutation, type MutationCtx, query } from "./_generated/server.js";
+import {
+  type OnCompleteArgs,
+  retryBehavior,
+  type RunResult,
+  vOnCompleteFnContext,
+  vResult,
```
[the remark this answers is not in the document]
Comment (sethconvex) [reply src/component/pool.ts]: Pool mode is not part of this PR — those files were in a separate branch. The comments on pool.ts are not applicable here.

File: src/component/batch.test.ts [hunk 3]
```diff
@@ -0,0 +1,665 @@
...
+  let t: Awaited<ReturnType<typeof setupTest>>;
+
+  beforeEach(async () => {
+    vi.useFakeTimers();
+    t = await setupTest();
+  });
+
+  afterEach(() => {
+    vi.useRealTimers();
+  });
```
[the remark this answers is not in the document]
Comment (sethconvex) [reply src/component/batch.test.ts]: Tests pass without timer cleanup — each test gets an isolated environment from `convex-test`. The fake timers are scoped to the test function.

File: src/client/batch.ts [hunk 4]
```diff
@@ -0,0 +1,750 @@
...
+  private options: BatchWorkpoolOptions;
+  private registry = new Map<string, HandlerFn>();
+  private executorFnRef: FunctionReference<
+    "action",
+    "internal"
+  > | null = null;
+  private cachedBatchConfig:
+    | { executorHandle: string; maxWorkers: number; claimTimeoutMs: number }
+    | undefined;
+  private configSentThisTx = false;
```
[the remark this answers is not in the document]
Comment (sethconvex) [reply src/client/batch.ts]: Tested WeakSet approach (resetting per-ctx) in a 40K-task benchmark. It caused 9,248 failures: every pipeline callback gets a fresh ctx, so every `batch.enqueue()` re-sent batchConfig → 80K `_maybeStartExecutors` mutations all hitting the batchConfig singleton → massive OCC storm. The boolean is intentionally instance-level to avoid this. After initial configuration on first enqueue, the executor handle and config don't change, so re-sending is unnecessary.

File: src/client/pool.ts [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+   * Generate the executor action. The user must export this from their
+   * convex/ directory:
+   *
+   * ```ts
+   * export const executor = pool.executor();
+   * pool.setExecutorRef(internal.pool.executor);
+   * ```
+   */
+  executor(): RegisteredAction<"internal", Record<string, never>, void> {
+    const self = this;
```
[the remark this answers is not in the document]
Comment (sethconvex) [reply src/client/pool.ts]: Pool mode files are not part of this PR — they were in a separate branch.

File: src/client/batch.ts [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+              self.component.pool.countPending,
+              {},
+            );
+            if (pending === 0) break;
+            await new Promise((r) => setTimeout(r, pollInterval));
+            continue;
+          }
+
+          // Wait for at least one task to finish before looping
+          await Promise.race([...inFlight.values()]);
```
[the remark this answers is not in the document]
Comment (sethconvex) [reply src/client/batch.ts]: Fixed — the executor loop uses `Promise.allSettled` for flush operations and catches errors at each level. The `flushBuffers` function retries transient errors and logs non-transient ones. Uncaught rejections are prevented by the `try/finally` block that awaits all in-flight work before exiting.

File: src/component/batch.ts [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+
+  // Start executors up to maxWorkers
+  if (config.activeExecutors < config.maxWorkers) {
+    const handle = config.executorHandle as FunctionHandle<"action">;
+    await ctx.db.patch(config._id, {
+      activeExecutors: config.activeExecutors + 1,
+    });
+    await ctx.scheduler.runAfter(0, handle, {});
+  }
+}
```
[the remark this answers is not in the document]
Comment (sethconvex) [reply src/component/batch.ts]: Fixed in a later commit — `ensureExecutors` now starts ALL missing slots in a single invocation (loops 0..maxWorkers-1, schedules for any missing).

File: example/convex/batchActions.ts [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+            role: "system",
+            content:
+              "Translate the following to Spanish. Return only the translation.",
+          },
+          { role: "user", content: sentence },
+        ],
+      }),
+    });
+    const data = await response.json();
+    return data.choices[0].message.content;
```
[the remark this answers is not in the document]
Comment (sethconvex) [reply example/convex/batchActions.ts]: Example code — intentionally kept simple. Users will add error handling appropriate for their use case.

File: src/client/batch.ts [hunk 9]
```diff
@@ -0,0 +1,467 @@
...
+        // Notify component that this executor is done
+        await ctx.runMutation(component.batch.executorDone, {
+          startMore: remaining > 0,
+        });
+
+        // Also self-continue for faster restart if we have a ref
+        const executorRef = getExecutorRef();
+        if (remaining > 0 && executorRef) {
+          await ctx.scheduler.runAfter(0, executorRef, {});
+        }
```
[the remark this answers is not in the document]
Comment (sethconvex) [reply src/client/batch.ts]: Fixed in commit 1ef10aa — eliminated double-scheduling. The executor no longer self-schedules; only `executorDone(startMore=true)` triggers replacement executors.

File: src/client/batch.ts [hunk 10]
```diff
@@ -0,0 +1,509 @@
...
+          {},
+        );
+
+        // Notify component that this executor is done.
+        // If there's remaining work, executorDone calls ensureExecutors
+        // to schedule a replacement — no self-scheduling needed.
+        await ctx.runMutation(component.batch.executorDone, {
+          startMore: remaining > 0,
+        });
+      },
```
[the remark this answers is not in the document]
Comment (sethconvex) [reply src/client/batch.ts]: Fixed in commit 1ef10aa — `executorDone` is now called inside the `finally` block with retry logic (10 retries, exponential backoff capped at 5s).

File: src/component/batch.ts [hunk 11]
```diff
@@ -0,0 +1,832 @@
...
+    for (const item of items) {
+      try {
+        await ctx.runMutation(
+          item.fnHandle as FunctionHandle<"mutation">,
+          {
+            workId: item.workId,
+            context: item.context,
+            result: item.result,
+          },
+        );
```
Comment (ianmacartney) [inline src/component/batch.ts]: note: this means each of them will share read/write bandwidth & OCC dependencies - so if any of them conflict it all runs again. e.g. a batch size of 100 limits each one to 160kb of reads/ 80kb of writes.
Might be better to have a single onComplete that expects to take an array, so they could share common reads, defer more processing via scheduler dynamically, etc. ?

**Review by ianmacartney — CHANGES_REQUESTED** [review]
Super cool model, and I think a compelling one for many tasks (especially doing pure fetch calls).
Given that it's a totally different implementation, let's make it a different component.
The big concerns are that it would blow through all available query/mutation capacity, starving other work (though more & more these "background" tasks are in the interaction path for users, so they aren't pure deferrable async jobs). 
In particular, having a batch job that calls runQuery / runMutation just after calling an LLM, I suspect would tank performance.

There's also a lot of places where it could easily exceed read/write bandwidth. I have a PR up adding a scenario test with large args, and was going to add one for large onComplete context & large return values. I'd suggest evaluating those to see where the edges are. Especially places like grabbing tasks / completing tasks - if they hit a batch that's too big, it grinds the whole thing to a halt
